File: wechat-ocr/monitor.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <stdarg.h>
#include <sys/stat.h>
#include <signal.h>

#define INTERVAL 300          // 5 分钟
#define LOG_FILE "/tmp/wechat_monitor.log"
#define TMP_DIR "/tmp/wechat_monitor"
#define SHELL_CMD_MAX 512

// 日志写入
static void log_msg(const char *fmt, ...) {
    char buf[256];
    va_list args;
    va_start(args, fmt);
    vsnprintf(buf, sizeof(buf), fmt, args);
    va_end(args);

    time_t t = time(NULL);
    struct tm *tm = localtime(&t);
    char ts[32];
    strftime(ts, sizeof(ts), "%H:%M:%S", tm);

    FILE *f = fopen(LOG_FILE, "a");
    if (f) {
        fprintf(f, "[%s] %s\n", ts, buf);
        fclose(f);
    }
    printf("[%s] %s\n", ts, buf);
    fflush(stdout);
}

// 执行 shell 命令
static int run_cmd(const char *fmt, ...) {
    char cmd[SHELL_CMD_MAX];
    va_list args;
    va_start(args, fmt);
    vsnprintf(cmd, sizeof(cmd), fmt, args);
    va_end(args);
    return system(cmd);
}

// 检测未读红点，返回红点数（-1 表示失败）
static int check_unread() {
    // 1. 激活微信
    run_cmd("xdotool search --name 微信 windowactivate 2>/dev/null");
    usleep(500000);

    // 2. 获取窗口位置
    char geo_file[128];
    snprintf(geo_file, sizeof(geo_file), "%s/geo.txt", TMP_DIR);
    run_cmd("xdotool getactivewindow getwindowgeometry > %s 2>/dev/null", geo_file);

    FILE *f = fopen(geo_file, "r");
    if (!f) return -1;
    char line[256];
    int wx = 0, wy = 0, ww = 0, wh = 0;
    while (fgets(line, sizeof(line), f)) {
        if (sscanf(line, " Position: %d,%d", &wx, &wy) == 2) continue;
        if (sscanf(line, " Geometry: %dx%d", &ww, &wh) == 2) continue;
    }
    fclose(f);
    if (ww == 0 || wh == 0) return -1;

    // 3. 截图
    char shot[128];
    snprintf(shot, sizeof(shot), "%s/wechat.png", TMP_DIR);
    run_cmd("import -window root -crop %dx%d+%d+%d '%s' 2>/dev/null", ww, wh, wx, wy, shot);

    // 4. 生成红色掩码
    run_cmd("convert '%s' +repage -fx '(r>g*2.2&&r>b*2.2&&r>0.7&&g<0.5)?1:0' '%s/_red.png' 2>/dev/null", shot, TMP_DIR);

    // 5. 连通分量分析
    char cc_file[128];
    snprintf(cc_file, sizeof(cc_file), "%s/cc.txt", TMP_DIR);
    run_cmd("convert '%s/_red.png' -define connected-components:verbose=true -connected-components 4 /dev/null 2>/dev/null | grep -v '0:\\|bgcolor\\|id:' > '%s'", TMP_DIR, cc_file);

    // 6. 统计符合条件的红点
    f = fopen(cc_file, "r");
    if (!f) return 0;

    int count = 0;
    while (fgets(line, sizeof(line), f)) {
        int id, w, h, x, y, px;
        if (sscanf(line, "%d: %dx%d+%d+%d %*f,%*f %d", &id, &w, &h, &x, &y, &px) >= 5) {
            // 过滤：大小 5~20px，左侧（x < 窗口宽*0.4），有效像素 >= 20
            if (w >= 5 && h >= 5 && w <= 20 && h <= 20 && x < ww * 0.4 && px >= 20) {
                count++;
            }
        }
    }
    fclose(f);
    return count;
}

// 信号处理
static volatile int keep_running = 1;
static void handle_signal(int sig) {
    keep_running = 0;
}

int main(int argc, char *argv[]) {
    int once = (argc > 1 && strcmp(argv[1], "--once") == 0);

    // 创建临时目录
    mkdir(TMP_DIR, 0755);

    // 守护进程模式
    if (!once) {
        pid_t pid = fork();
        if (pid < 0) { perror("fork"); return 1; }
        if (pid > 0) {
            printf("[微信监控] 守护进程已启动 (PID: %d)\n", pid);
            printf("日志: %s\n", LOG_FILE);
            return 0;  // 父进程退出
        }
        // 子进程：新会话
        setsid();
        chdir("/");
        // 关闭标准 I/O
        fclose(stdin); fclose(stdout); fclose(stderr);
        // 打开日志作为 stdout/stderr
        freopen(LOG_FILE, "a", stdout);
        freopen(LOG_FILE, "a", stderr);
        setvbuf(stdout, NULL, _IONBF, 0);
    }

    // 信号处理
    signal(SIGINT, handle_signal);
    signal(SIGTERM, handle_signal);

    log_msg("微信监控启动%s", once ? "（单次模式）" : "（每5分钟检测）");

    int prev_count = 0;
    while (keep_running) {
        int count = check_unread();
        if (count >= 0) {
            if (count > 0) {
                log_msg("[%d 条未读]", count);
                if (count != prev_count) {
                    prev_count = count;
                    log_msg("!! 有新消息 !!");
                }
            } else {
                log_msg("[无新消息]");
                prev_count = 0;
            }
        } else {
            log_msg("[错误] 检测失败");
        }
        if (once) break;

        // 等待 INTERVAL 秒（每 5 秒检查一次信号）
        for (int i = 0; i < INTERVAL && keep_running; i++) {
            sleep(1);
        }
    }

    log_msg("监控停止");
    return 0;
}

File: wechat-ocr/tests/monitor.lua
#!/usr/bin/env luajit
-- WeChat 消息监控守护进程
-- 每 5 分钟检测第二列是否有红点（未读消息）
-- 用法: luajit monitor.lua &          # 后台运行
--       luajit monitor.lua --once     # 只检测一次

local ffi = require("ffi")
ffi.cdef[[int usleep(unsigned int);]]

----------------------------------------
-- 配置
----------------------------------------
local CHECK_INTERVAL = 300  -- 5 分钟（秒）
local SCREENSHOT_DIR = "/tmp/wechat_monitor"

-- 颜色输出
local RED = "\27[31m"
local GREEN = "\27[32m"
local RESET = "\27[0m"

----------------------------------------
-- 检测红点（用连通分量分析）
----------------------------------------
local function check_red_dots()
    -- 截图微信窗口
    os.execute("mkdir -p " .. SCREENSHOT_DIR)
    os.execute("xdotool search --name 微信 windowactivate 2>/dev/null")
    ffi.C.usleep(500000)

    local geo = io.popen("xdotool getactivewindow getwindowgeometry 2>/dev/null"):read("*a")
    local win_x = tonumber(geo:match("Position: (%d+)"))
    local win_y = tonumber(geo:match(",(%d+)"))
    local win_w = tonumber(geo:match("Geometry: (%d+)"))
    local win_h = tonumber(geo:match("x(%d+)"))
    if not win_w then return -1, "无法获取微信窗口" end

    local shot = SCREENSHOT_DIR .. "/wechat.png"
    os.execute(string.format("import -window root -crop %dx%d+%d+%d '%s' 2>/dev/null", win_w, win_h, win_x, win_y, shot))

    -- 用 ImageMagick 分析红色区域
    -- R > G*2.2, R > B*2.2, R > 0.7, G < 0.5
    os.execute(string.format(
        "convert '%s' +repage -fx '(r>g*2.2&&r>b*2.2&&r>0.7&&g<0.5)?1:0' '/tmp/_mon_red.png' 2>/dev/null", shot))

    -- 连通分量分析
    local pipe = io.popen(string.format(
        "convert '/tmp/_mon_red.png' -define connected-components:verbose=true -connected-components 4 /dev/null 2>/dev/null | grep -v '0:\\|bgcolor\\|id:'"))
    local output = pipe:read("*a"); pipe:close()

    local count = 0
    for line in output:gmatch("[^\n]+") do
        local id, w, h, x, y, px = line:match("(%d+):%s*(%d+)x(%d+)%+(%d+)%+(%d+)%s+[%d.]+,[%d.]+%s+(%d+)")
        if w and x and y and px then
            local nw, nh, nx, ny, np = tonumber(w), tonumber(h), tonumber(x), tonumber(y), tonumber(px)
            if nw >= 5 and nh >= 5 and nw <= 20 and nh <= 20 and nx < win_w * 0.4 and np >= 20 then
                count = count + 1
            end
        end
    end
    return count, nil
end

----------------------------------------
-- 主循环
----------------------------------------
local once = (arg and arg[1] == "--once")

io.write(string.format("%s [微信监控] 启动%s\n", os.date("%H:%M:%S"), once and "（单次模式）" or "（每5分钟检测）"))
io.flush()

local prev_count = 0

while true do
    local ok, msg = pcall(check_red_dots)
    if ok then
        local count = msg
        local ts = os.date("%H:%M:%S")
        if count > 0 then
            io.write(string.format("%s %s[%d 条未读]%s\n", ts, RED, count, RESET))
            if count ~= prev_count then
                prev_count = count
                -- 可以在这里添加通知逻辑，如播放声音、发消息等
            end
        else
            io.write(string.format("%s %s[无新消息]%s\n", ts, GREEN, RESET))
            prev_count = 0
        end
    else
        io.write(string.format("%s [错误] %s\n", os.date("%H:%M:%S"), tostring(msg)))
    end
    io.flush()

    if once then break end
    -- 等待间隔（每 10 秒检查一次是否到时间）
    for i = 1, CHECK_INTERVAL do
        ffi.C.usleep(1000000)
        -- 可以在这里添加信号处理
    end
end

io.write("监控结束\n")
